launchtemplate docs (#773)
awslabs/karpenter: website/content/en/docs/cloud-providers/AWS/launch-templates.md

---
title: "Launch Templates and Custom Images"
linkTitle: "Launch Templates"
weight: 80
---

By default, Karpenter generates launch templates that use [EKS Optimized AMI](https://docs.aws.amazon.com/eks/latest/userguide/eks-optimized-ami.html) for nodes. Often, users need to customize the node image to integrate with existing infrastructure or meet compliance requirements. Karpenter supports custom node images through Launch Templates. If you need to customize the node, then you need a custom launch template. 

> Note: By customizing the image, you are taking responsibility for maintianing the image, including security updates. In the default configuration, Karpenter will use the latest version of the EKS optimized AMI, which is maintained by AWS. 

Reviewer: Typo: maintianing

Author: i miss msword sometimes

## Introduction

Karpenter follows existing AWS patterns for customizing the base image of
instances. More specifically, Karpenter uses [EC2 launch templates](https://docs.aws.amazon.com/autoscaling/ec2/userguide/LaunchTemplates.html). Launch
templates may specify many values. The pivotal value is the base image (AMI).
Launch templates further specify many different parameters related to networking, authorization, instance type, and more. 

**Karpenter only implementes a subset of launch template fields, and some fields should not be set.**  

This guide describes requirements for using launch templates with Karpenter, and later an example procedure.

## Launch Template Requirements

The Launch Template resource includes a large number of fields. AWS accepts launch templates with any subset of these fields defined.

> Certain fields are obviously critical, such as AMI and User Data. Some fields are useful for particular workloads, such as storage and IAM Instance Profile. 

Reviewer: Nit: probably wouldn't use the word "obviously" - if it was obvious, they probably wouldn't be reading your doc :)

Author: eh im going to keep it as is, certain fields are more obviously useful (image), than others (security groups).

> Finally, **the majority of Launch Template fields should not be set** (or will have no effect), such as network interfaces and instance type. 

Reviewer: Feels a bit ambiguous, should we just say:

```suggestion
Finally, **the majority of Launch Template fields should not be set** (or will have no effect). If it's not documented below, you should assume that the field should not be set.
```

Author: I think part of your message got lost?

## Important Fields

When creating a custom launch template, the AMI and User Data are the defining characteristics. Instance Profile (IAM Role) and Security Group (firewall rules) are also important for Karpenter. 

### AMI

> AMI (Amazon Machine Image), is the base image/VM for a launch template.

Reviewer: For those new to AWS, is it worth noting that the AMI is regional? (for that matter, I wonder if we should note that LTs are regional?)

Author: this is a good point.

[Review the instructions for importing a VM to AWS.](https://docs.aws.amazon.com/vm-import/latest/userguide/vmimport-image-import.html) Note the AMI id generated by this process, such as, 
`ami-074cce78125f09d61`.

### User Data - Autoconfigure

Importantly, the AMI must support automatically connecting to a cluster based
on "user data", or a base64 encoded string passed to the instance at startup.
The syntax and purpose of the user data varies between images. The Karpenter

> default OS, Amazon Linux 2 (AL2), accepts shell scripts (bash commands). 

Reviewer: Does it exclusively accept bash? If so, should we say:

```suggestion
default OS, Amazon Linux 2 (AL2), accepts bash scripts. 
```

else:

```suggestion
default OS, Amazon Linux 2 (AL2), accepts shell scripts (e.g. bash commands). 
```

[AWS calls data passed to an instance at launch time "user

> data".](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/user-data.html#user-data-shell-scripts)

Reviewer: Feels like this should go before the previous paragraph. Either the reader understood this from the previous paragraph, or they didn't and that previous paragraph confused them. IMHO, either we should remove this or put it first.

Author: I'm happy with the first sentence.

> Importantly, the AMI must support automatically connecting to a cluster based
on "user data", or a base64 encoded string passed to the instance at startup.

There is no graceful way to introduce these confusing terms.

[AWS calls data passed to an instance at launch time "user
data".](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/user-data.html#user-data-shell-scripts)

In the default configuration, Karpenter uses an EKS optimized version of AL2 and passes the hostname of the Kubernetes API server, and a certificate. The EKS Optimized AMI includes a `bootstrap.sh` script which connects the instance to the cluster, based on the passed data. 

Alternatively, you may reference AWS's [`bootstrap.sh`
file](https://github.com/awslabs/amazon-eks-ami/blob/master/files/bootstrap.sh)
when building a custom base image. 

```
#!/bin/bash
/etc/eks/bootstrap.sh <my-cluster-name> \
--kubelet-extra-args <'--max-pods=40'> \
--b64-cluster-ca <certificateAuthority> \

> --apiserver-endpoint <endpoint> 

Reviewer 2: need `\` on this line and the next

```

> Note, you must populate this command with live values. Karpenter will

Reviewer: Nit: "live" to me would imply dynamic. Do you mean "real, static" values?

not change the user data in the launch template. 

> Encode using yaml function `!Base64` yaml function or `cat userdata.sh | base64 > userdata-encoded.txt` shell command. 

Reviewer: Nit: 
```suggestion
You can encode your user data using yaml function `!Base64` yaml function or shell command `cat userdata.sh | base64 > userdata-encoded.txt` . 
```

**Bootstrap Script Parameters**

The sample bootstrap script requires information to join the cluster.

These values may be found using:
```

> aws eks describe-cluster --name MyKarpenterCluster 

Reviewer: Nit: following your earlier convention for substitutions, should this be:

```suggestion
aws eks describe-cluster --name <MyKarpenterCluster>

```

```

> **Kubelet Arguments**

Reviewer: This section doesn't talk about other kubelet arguments. Is the implication that you _can_ configure other parameters?

Author: yes, we may add to this section in the future.

**Kubelet Arguments**

> Specifying max-pods can break Karpenter's binpacking logic (it has no way to know what this setting is). If karpenter attempts to pack more than this number of pods, the instance may be oversized, and additional pods will reschedule.

Reviewer 2: `s/karpenter/Karpenter/`?

## Situational Fields 

Configure these values in response to a particular use case, such as nodes interacting with another AWS service, or using EBS storage on the node. 

### Instance Profile - IAM

> The launch template must include an "instance profile" -- a set of IAM roles. 

Reviewer 2: can you say `an IAM role`

Author: sorry this got dropped?

> The instance profile must include *at least* the permissions of the default Karpenter node instance profile. See the default role, `KarpenterNodeRole`, in the full example below for more information. 

Reviewer: Can we use [an anchor link](https://gist.github.com/asabaylus/3071099)?

```suggestion
The instance profile must include *at least* the permissions of the default Karpenter node instance profile. See the default role, `KarpenterNodeRole`, in the [full example below](#cloudformation) for more information. 

```

See also, [the managed policy "AmazonEKSWorkerNodePolicy"](https://docs.aws.amazon.com/eks/latest/userguide/security-iam-awsmanpol.html#security-iam-awsmanpol-AmazonEKSWorkerNodePolicy) which includes permission to describe clusters and subnets.

### Storage

Karpenter expects nothing of node storage. Configure as needed for your base
image.

### Security Groups - Firewall

> The launch template must include a security group (i.e., instance firewall rules) and the security group must be associated with the virtual private cloud (VPC) of the EKS cluster.

Reviewer: Does this mean all custom LTs for Karpenter _must_ have a security group? Or is this just a general comment? If it's for all custom LTs, should this get promoted to the "Important Fields" section? If it's not, can you call out that Karpenter will set this up if you don't?

Author: i'll ask about this in slack.

Author: I just diy'd it. Doesn't need one. Will use the VPC default. Updated desc.

> The security group must permit communication with EKS control plane. Outbound access should be permitted for at least: HTTPS on port 443, DNS (UDP and TCP) on port 53, and your subnet's network access control list (network ACL). 

Reviewer: Nit: consider a bulleted list

```suggestion
The security group must permit communication with EKS control plane. Outbound access should be permitted for at least: 
- HTTPS on port 443
- DNS (UDP and TCP) on port 53
- Your subnet's network access control list (network ACL). 

```

## Fields with Undefined Behavior

Resources referenced by these fields are controlled by EKS/Karpenter, and not the launch template. 

### Instance Type

The instance type should not be specified in the launch template. Karpenter
will determine the launch template at run time. 

### Network Interfaces

> The [AWS CNI](https://docs.aws.amazon.com/eks/latest/userguide/pod-networking.html) will configure the network interfaces. Do not configure network instances in the launch template.

Reviewer: Isn't there an alternative to using the CNI? (I can't recall details) Or would we not want to document that?

Author: I do not want to document that. If you use a non-standard CNI, it's your responsibilty.

The [AWS CNI](https://docs.aws.amazon.com/eks/latest/userguide/pod-networking.html) will configure the network interfaces. Do not configure network instances in the launch template.

## Creating the Launch Template

Launch Templates may be created via the web console, the AWS CLI, or
CloudFormation. 

### CloudFormation

The procedure, in summary, is to:
1. [Create an AMI as described in the EC2 documentation.](https://docs.aws.amazon.com/vm-import/latest/userguide/vmimport-image-import.html)
2. Write a EC2 Launch Template specification including the AMI.
3. Push the specification to AWS with CloudFormation.
4. Update the Provisioner CRD to specify the new Launch Template.

An example yaml cloudformation definition of a launch template for Karpenter is
provided below. 

> Cloudformation yaml is suited for the moderately high configuration density of

Reviewer 2: `s/Cloudformation/CloudFormation/`

        - Ref: "KarpenterNodeRole"
  # create role with basic permissions for EKS node
  KarpenterNodeRole:
    Type: "AWS::IAM::Role"
    Properties:
      RoleName: "KarpenterNodeRole"
      Path: /
      AssumeRolePolicyDocument:
        Version: "2012-10-17"
        Statement:
          - Effect: Allow
            Principal:
              Service:
                !Sub "ec2.${AWS::URLSuffix}"
            Action:
              - "sts:AssumeRole"
      ManagedPolicyArns:
        - !Sub "arn:${AWS::Partition}:iam::aws:policy/AmazonEKSWorkerNodePolicy"
        - !Sub "arn:${AWS::Partition}:iam::aws:policy/AmazonEKS_CNI_Policy"
        - !Sub "arn:${AWS::Partition}:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly"
        - !Sub "arn:${AWS::Partition}:iam::aws:policy/AmazonSSMManagedInstanceCore"
  MyLaunchTemplate:
    Type: AWS::EC2::LaunchTemplate
    Properties: 
      LaunchTemplateData: 
        IamInstanceProfile:
          # Get ARN of InstanceProfile defined above
          Arn: !GetAtt
            - KarpenterInstanceProfile
            - Arn
        ImageId: ami-074cce78125f09d61
        # UserData is Base64 Encoded
        UserData: !Base64 > 
            #!/bin/bash

>             /etc/eks/bootstrap.sh <my-cluster-name> \

Reviewer: If you're going to use actual values elsewhere in this example, why not also in the UserData?

Author: this is a good point. i'll update it to use actual values.

            --kubelet-extra-args <'--max-pods=40'> \
            --b64-cluster-ca <certificateAuthority> \
            --apiserver-endpoint <endpoint> 
            --dns-cluster-ip <serviceIpv4Cidr>
            --use-max-pods false
        BlockDeviceMappings: 
          - Ebs:
              VolumeSize: 80
              VolumeType: gp3
            DeviceName: /dev/xvda
        # The SecurityGroup must be associated with the cluster VPC
        SecurityGroupIds:
          - sg-a69adfdb
      LaunchTemplateName: KarpenterCustomLaunchTemplate
```

Create the Launch Template by uploading the CloudFormation yaml file. The
sample yaml creates an IAM Object (InstanceProfile), so `--capabilities
CAPABILITY_NAMED_IAM` must be indicated.

```
aws cloudformation create-stack \
  --stack-name KarpenterLaunchTemplateStack \
  --template-body file://$(pwd)/lt-cfn-demo.yaml \
  --capabilities CAPABILITY_NAMED_IAM
```

> ### Define LaunchTemplate for Provisioner

Reviewer: This is an actual question I have: what if you support different architectures? Can you supply different custom LTs to support the different architectures?

Author: This is an onging question that (as far as I know) hasn't been resolved by engineering.

Author: I've decided supporting multiple archs (arches?) is out of scope for this PR.

---

> title: "Launch Templates and Custom Images"

Reviewer 3: I'd just call this `LaunchTemplates`

Author: I'd like to offer a bread crumb to non-AWS people. Do you think it's intuitive that "custom node image" is covered by "launch template"?

linkTitle: "Launch Templates"
weight: 80
---

> By default, Karpenter uses the EKS Optimized version of Amazon Linux 2 (AL2) for nodes. Often, users need to customize the node image to integrate with existing infrastructure or meet compliance requirements. Karpenter supports custom node images through Launch Templates. If you need to customize the node, then you need a custom launch template. 

Reviewer 3: ```suggestion
By default, Karpenter generates launch templates that use [EKS Optimized AMI](https://docs.aws.amazon.com/eks/latest/userguide/eks-optimized-ami.html) for nodes. Often, users need to customize the node image to integrate with existing infrastructure or meet compliance requirements. Karpenter supports custom node images through Launch Templates. If you need to customize the node, then you need a custom launch template. 

```

## Introduction

Karpenter follows existing AWS patterns for customizing the base image of
instances. More specifically, Karpenter uses [EC2 launch templates](https://docs.aws.amazon.com/autoscaling/ec2/userguide/LaunchTemplates.html). Launch
templates may specify many values. The pivotal value is the base image (AMI).
Launch templates further specify many parameters related to networking,

> authorization, instance type, and more.  

Reviewer 2: remove `instance type` because karpenter overrides the LT instance type, so it will never be used even if a customer specifies it in their custom LT. 

EDIT: I see you say never to specify the instance type in the LT Requirements section. I kind of still think it should be removed here to not confuse anyone. But I'll leave that up to you.

This guide describes requirements for using launch templates with Karpenter, and later an example procedure.

## Launch Template Requirements

The Launch Template resource includes a large number of fields. AWS accepts launch templates with any subset of these fields defined.

Certain fields are obviously critical, such as AMI and User Data. Some fields are useful for particular workloads, such as storage and IAM Instance Profile. 

Finally, **the majority of Launch Template fields should not be set** (or will have no effect), such as network interfaces and instance type. 

## Important Fields

> When creating a custom launch template, the AMI and User Data are the defining characteristics. 

Reviewer 2: For Karpenter, the Instance Profile and Security Group are also critical.

when building a custom base image. 

```
#!/bin/bash
/etc/eks/bootstrap.sh <my-cluster-name> \
--kubelet-extra-args <'--max-pods=40'> \
--b64-cluster-ca <certificateAuthority> \
--apiserver-endpoint <endpoint> 
--dns-cluster-ip <serviceIpv4Cidr>
--use-max-pods false
```

Note, you must populate this command with live values. Karpenter will
not change the user data in the launch template. 

Encode using yaml function `!Base64` yaml function or `cat userdata.sh | base64 > userdata-encoded.txt` shell command. 

**Bootstrap Script Parameters**

The sample bootstrap script requires information to join the cluster.

These values may be found using:
```
aws eks describe-cluster --name MyKarpenterCluster 
```

**Kubelet Arguments**

Specifying max-pods can break Karpenter's binpacking logic (it has no way to know what this setting is). If karpenter attempts to pack more than this number of pods, the instance may be oversized, and additional pods will reschedule.

## Situational Fields 

Configure these values in response to a particular use case, such as nodes interacting with another AWS service, or using EBS storage on the node. 

### Instance Profile - IAM

> The launch template must include an "instance profile" -- a set of IAM roles. 

Reviewer 2: `a set of IAM policies` instance profiles are 1:1 w/ roles.

Author: <img width="529" alt="Screen Shot 2021-11-09 at 9 14 36 AM" src="https://user-images.githubusercontent.com/5782953/140950977-d55bbdb8-a6f7-43df-832d-909cfefd6af9.png">

To me, this looks like "list of roles" but perhaps I'm interpreting it wrong?

Reviewer 2: Very confusing, but InstanceProfile's can only have 1 role:

https://awscli.amazonaws.com/v2/documentation/api/latest/reference/iam/add-role-to-instance-profile.html#description

```
Adds the specified IAM role to the specified instance profile. An instance profile can contain only one role, and this quota cannot be increased
```

Author: *screams*

The instance profile must include all the permissions of the default Karpenter

> node instance profile. For example, permission to run containers and manage

Reviewer 2: I'm not sure what you mean by `run containers`? The Instance profile shouldn't have anything special to "run containers", unless I'm misunderstanding.

Author: I'm going to replace this with a link to "AmazonEKSWorkerNodePolicy"

networking. See the default role, `KarpenterNodeRole`, in the full example
below for more information. 

### Storage

Karpenter expects nothing of node storage. Configure as needed for your base
image.

### Security Groups - Firewall

> EKS configures security groups (i.e., instance firewall rules) automatically. 

Reviewer 2: This seems too general and maybe not true (at least in all cases that would matter to a customer reading this specific doc). Do you mean that EKS MNG's configure security groups, or eksctl for self-managed?   

For the case of a user providing a custom LT to Karpenter, they would need to supply the security group to the LT.

Author: I was just wrong.

However, you may manually specify a security group. The security group must
permit communication with EKS control plane. Outbound access should be
permitted for at least: HTTPS on port 443, DNS (UDP and TCP) on port 53, and
your subnet's network access control list (network ACL). 

The security group must be associated with the virtual private cloud (VPC) of
the EKS cluster.

## Fields with Undefined Behavior

> These resources referenced by these fields are controlled by EKS/Karpenter, and not the launch template. 

Reviewer 2: `These resources referenced by these`. I think you can just drop the first `These`

### Instance Type

The instance type should not be specified in the launch template. Karpenter
will determine the launch template at run time. 

### Network Interfaces

> EKS will configure the network interfaces. Do not configure network instances

Reviewer 2: The CNI will configure the network interfaces, not necessarily EKS.

    Properties:
      InstanceProfileName: "KarpenterNodeInstanceProfile"
      Path: "/"
      Roles:
        - Ref: "KarpenterNodeRole"
  # create role with basic permissions for EKS node
  KarpenterNodeRole:
    Type: "AWS::IAM::Role"
    Properties:
      RoleName: "KarpenterNodeRole"
      Path: /
      AssumeRolePolicyDocument:
        Version: "2012-10-17"
        Statement:
          - Effect: Allow
            Principal:
              Service:
                !Sub "ec2.${AWS::URLSuffix}"
            Action:
              - "sts:AssumeRole"
      ManagedPolicyArns:
        - !Sub "arn:${AWS::Partition}:iam::aws:policy/AmazonEKSWorkerNodePolicy"
        - !Sub "arn:${AWS::Partition}:iam::aws:policy/AmazonEKS_CNI_Policy"
        - !Sub "arn:${AWS::Partition}:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly"
        - !Sub "arn:${AWS::Partition}:iam::aws:policy/AmazonSSMManagedInstanceCore"
  MyLaunchTemplate:
    Type: AWS::EC2::LaunchTemplate
    Properties: 
      LaunchTemplateData: 
        IamInstanceProfile:
          # Get ARN of InstanceProfile defined above
          Arn: !GetAtt
            - KarpenterInstanceProfile
            - Arn

>         ImageId: ami-074cce78125f09d61

Reviewer 2: is this a public AMI? If so, do you need to specify the correct region for this to run? 

It might be better to use `<my-ami-id>` so the user knows to fill it in.

Author: As a personal preference, I like seeing what the actual value looks like. I have the list of values that must be replaced above.

Reviewer 2: fair enough :)

        # UserData is Base64 Encoded
        UserData: !Base64 > 
            #!/bin/bash
            /etc/eks/bootstrap.sh <my-cluster-name> \
            --kubelet-extra-args <'--max-pods=40'> \
            --b64-cluster-ca <certificateAuthority> \
            --apiserver-endpoint <endpoint> 
            --dns-cluster-ip <serviceIpv4Cidr>
            --use-max-pods false
        BlockDeviceMappings: 
          - Ebs:
              VolumeSize: 80
              VolumeType: gp3
            DeviceName: /dev/xvda
        # The SecurityGroup must be associated with the cluster VPC
        SecurityGroupIds:

>           - sg-a69adfdb

Reviewer 2: same here, maybe `<my-security-group-id>`

      LaunchTemplateName: KarpenterCustomLaunchTemplate
```

Create the Launch Template by uploading the CloudFormation yaml file. The
sample yaml creates an IAM Object (InstanceProfile), so `--capabilities
CAPABILITY_NAMED_IAM` must be indicated.

```
aws cloudformation create-stack \
  --stack-name KarpenterLaunchTemplateStack \

>   --template-body file:///Users/gcline/Desktop/lt-cfn-demo.yaml \

Reviewer 2: can you remove your `gcline` path? :D  

Maybe just replace with `file://$(pwd)/lt-cfn-demo.yaml`
